diff -Napur linux-linux4microchip-2025.04-old/drivers/net/phy/micrel.c linux-linux4microchip-2025.04-new/drivers/net/phy/micrel.c
--- linux-linux4microchip-2025.04-old/drivers/net/phy/micrel.c	2025-04-25 12:11:41.000000000 -0700
+++ linux-linux4microchip-2025.04-new/drivers/net/phy/micrel.c	2025-06-18 17:16:25.010344537 -0700
@@ -5669,6 +5669,8 @@ static struct phy_driver ksphy_driver[]
 	.name		= "Micrel KSZ87XX Switch",
 	/* PHY_BASIC_FEATURES */
 	.config_init	= kszphy_config_init,
+	.config_intr	= kszphy_config_intr,
+	.handle_interrupt = kszphy_handle_interrupt,
 	.match_phy_device = ksz8795_match_phy_device,
 	.suspend	= genphy_suspend,
 	.resume		= genphy_resume,
