diff -Napur linux-linux4microchip-2025.04-old/drivers/net/dsa/microchip/ksz8.c linux-linux4microchip-2025.04-new/drivers/net/dsa/microchip/ksz8.c
--- linux-linux4microchip-2025.04-old/drivers/net/dsa/microchip/ksz8.c	2025-06-19 15:55:16.237423118 -0700
+++ linux-linux4microchip-2025.04-new/drivers/net/dsa/microchip/ksz8.c	2025-07-29 13:59:08.120877365 -0700
@@ -371,6 +371,9 @@ static void ksz8863_r_mib_pkt(struct ksz
 	addr -= dev->info->reg_mib_cnt;
 	ctrl_addr = addr ? KSZ8863_MIB_PACKET_DROPPED_TX_0 :
 			   KSZ8863_MIB_PACKET_DROPPED_RX_0;
+	if (ksz_is_8895_family(dev) &&
+	    ctrl_addr == KSZ8863_MIB_PACKET_DROPPED_RX_0)
+		ctrl_addr = KSZ8895_MIB_PACKET_DROPPED_RX_0;
 	ctrl_addr += port;
 	ctrl_addr |= IND_ACC_TABLE(TABLE_MIB | TABLE_READ);
 
diff -Napur linux-linux4microchip-2025.04-old/drivers/net/dsa/microchip/ksz8_reg.h linux-linux4microchip-2025.04-new/drivers/net/dsa/microchip/ksz8_reg.h
--- linux-linux4microchip-2025.04-old/drivers/net/dsa/microchip/ksz8_reg.h	2025-04-25 12:11:41.000000000 -0700
+++ linux-linux4microchip-2025.04-new/drivers/net/dsa/microchip/ksz8_reg.h	2025-07-29 14:00:07.735572840 -0700
@@ -784,7 +784,9 @@
 #define KSZ8795_MIB_TOTAL_TX_1		0x105
 
 #define KSZ8863_MIB_PACKET_DROPPED_TX_0 0x100
-#define KSZ8863_MIB_PACKET_DROPPED_RX_0 0x105
+#define KSZ8863_MIB_PACKET_DROPPED_RX_0 0x103
+
+#define KSZ8895_MIB_PACKET_DROPPED_RX_0 0x105
 
 #define MIB_PACKET_DROPPED		0x0000FFFF
 
