diff -Napur linux-linux4microchip-2025.04-old/drivers/net/dsa/microchip/ksz_spi.c linux-linux4microchip-2025.04-old1/drivers/net/dsa/microchip/ksz_spi.c
--- linux-linux4microchip-2025.04-old/drivers/net/dsa/microchip/ksz_spi.c	2025-10-06 16:44:59.371530283 -0700
+++ linux-linux4microchip-2025.04-old1/drivers/net/dsa/microchip/ksz_spi.c	2025-10-06 16:12:08.429103762 -0700
@@ -184,12 +184,6 @@ static int ksz_spi_probe(struct spi_devi
 	if (spi->dev.platform_data)
 		dev->pdata = spi->dev.platform_data;
 
-	/* setup spi */
-	spi->mode = SPI_MODE_3;
-	ret = spi_setup(spi);
-	if (ret)
-		return ret;
-
 	dev->irq = spi->irq;
 
 	ret = ksz_switch_register(dev);
