diff -Napur linux-linux4microchip-2024.04-old/drivers/net/phy/phy_device.c linux-linux4microchip-2024.04-new/drivers/net/phy/phy_device.c
--- linux-linux4microchip-2024.04-old/drivers/net/phy/phy_device.c	2024-05-24 05:04:06.000000000 -0700
+++ linux-linux4microchip-2024.04-new/drivers/net/phy/phy_device.c	2026-04-27 19:21:31.589586643 -0700
@@ -357,8 +357,10 @@ static __maybe_unused int mdio_bus_phy_r
 	 * that something went wrong and we should most likely be using
 	 * MAC managed PM, but we are not.
 	 */
+#ifndef CONFIG_NET_DSA_MICROCHIP_KSZ_COMMON
 	WARN_ON(phydev->state != PHY_HALTED && phydev->state != PHY_READY &&
 		phydev->state != PHY_UP);
+#endif
 
 	ret = phy_init_hw(phydev);
 	if (ret < 0)
