diff -Napur linux-linux4microchip-2022.10-orig/drivers/ptp/ptp_clock.c linux-linux4microchip-2022.10/drivers/ptp/ptp_clock.c
--- linux-linux4microchip-2022.10-orig/drivers/ptp/ptp_clock.c	2022-11-09 04:48:42.000000000 -0800
+++ linux-linux4microchip-2022.10/drivers/ptp/ptp_clock.c	2024-03-01 02:12:09.599272081 -0800
@@ -147,6 +147,17 @@ static int ptp_clock_adjtime(struct posi
 		}
 	} else if (tx->modes == 0) {
 		tx->freq = ptp->dialed_frequency;
+#if defined(CONFIG_KSZ_PTP) || defined(CONFIG_LAN937X_PTP)
+		do {
+			s64 freq;
+
+			err = ops->gettime64(ops, NULL);
+			freq = err;
+			freq <<= 13;
+			freq = div64_s64(freq, 125);
+			tx->freq = freq;
+		} while (0);
+#endif
 		err = 0;
 	}
 
